utt/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,181 @@
+  handleSigJob(sigIds, {partial_sig.second}, job_timeout);
+}
+
+void SigProcessor::handleSigJob(std::vector<uint64_t> sigIds,
+                                const std::vector<libutt::api::types::Signature>& partial_sigs,
+                                uint64_t job_timeout) {
+  if (sigIds.size() != partial_sigs.size()) return;
+  for (size_t i = 0; i < sigIds.size(); i++) {
+    auto sigId = sigIds.at(i);
+    auto& partial_sig = partial_sigs.at(i);
+    auto& jobEntry = jobs_.get(sigId);
+    auto& validator = jobs_.getJobValidator(sigId);
+    std::vector<uint16_t> invalids;
+    for (auto& [id, psig] : jobEntry.partial_sigs) {
+      if (!validator((uint16_t)id, psig)) invalids.push_back((uint16_t)id);
+    }
+    for (auto id : invalids) jobEntry.partial_sigs.erase((uint32_t)id);
+
+    jobEntry.partial_sigs.emplace((uint32_t)repId_, partial_sig);
+    jobEntry.job_id = sigId;
+    jobEntry.start_time = std::chrono::steady_clock::now().time_since_epoch().count();
+    jobEntry.job_timeout_ms_ = job_timeout;
+    jobs_.put(sigId, jobEntry);
+    auto job_coordinator = getJobCoordinator(sigId);
+    uint16_t rep = repId_;
+    messages::PartialSigMsg msg(rep, partial_sig, sigId);
+    msgsCommunicator_->sendAsyncMessage((uint64_t)job_coordinator, msg.body(), msg.size());
+  }
+}
+SigProcessor::~SigProcessor() { timers_.cancel(timeout_handler_); }
+void SigProcessor::onReceivingNewPartialSig(uint64_t sigId,
+                                            uint16_t sig_source,
+                                            const libutt::api::types::Signature& partial_sig) {
+  SigJobEntry& entry = jobs_.get(sigId);
+  if (entry.partial_sigs.size() == threshold_) return;
+  if (jobs_.isRegistered(sigId)) {
+    auto& validator = jobs_.getJobValidator(sigId);
+    if (!validator(sig_source, partial_sig)) return;
+  }
+  entry.partial_sigs.emplace((uint32_t)sig_source, partial_sig);
+  if (entry.partial_sigs.size() == threshold_) {
+    auto sig = libutt::api::Utils::aggregateSigShares(n_, entry.partial_sigs);
+    auto requestSeqNum =
+        std::chrono::duration_cast<std::chrono::microseconds>(getMonotonicTime().time_since_epoch()).count();
+    std::vector<uint8_t> appClientReq = generateAppClientRequest(sigId, sig);
+    auto crm = new bftEngine::impl::ClientRequestMsg(repId_,
+                                                     0x0,
+                                                     requestSeqNum,
+                                                     (uint32_t)appClientReq.size(),
+                                                     (const char*)appClientReq.data(),
+                                                     60000,
+                                                     "new-utt-sig-" + std::to_string(sigId));
+    msgsCommunicator_->getIncomingMsgsStorage()->pushExternalMsg(std::unique_ptr<MessageBase>(crm));
+  }
+}
+void SigProcessor::onReceivingNewValidFullSig(uint64_t sigId) { jobs_.erase(sigId); }
```

Comment:
What calls this and when ?

Reply:
the user of sigProcessor call it when it gets through a consensus a message with a valid full signature (in the case of Ethereum, it will be called by the EVM on the execution step)